cpp/src/barretenberg/honk/transcript/transcript.hpp

```diff
@@ -0,0 +1,267 @@
+#pragma once
+
+#include "barretenberg/common/serialize.hpp"
+#include "barretenberg/crypto/pedersen/pedersen.hpp"
+#include "barretenberg/crypto/blake3s/blake3s.hpp"
+
+#include <array>
+#include <concepts>
+#include <cstddef>
+#include <cstdint>
+#include <span>
+#include <string>
+#include <utility>
+#include <vector>
+#include <map>
+
+namespace honk {
+
+class TranscriptManifest {
+    struct RoundData {
+        std::vector<std::string> challenge_label;
+        std::vector<std::pair<std::string, size_t>> entries;
+
+        bool operator==(const RoundData& other) const = default;
+    };
+
+    std::map<size_t, RoundData> manifest;
+
+  public:
+    void print()
+    {
+        for (auto& round : manifest) {
+            info("Round: ", round.first);
+            for (auto& label : round.second.challenge_label) {
+                info("\tchallenge: ", label);
+            }
+            for (auto& entry : round.second.entries) {
+                info("\telement (", entry.second, "): ", entry.first);
+            }
+        }
+    }
+
+    template <typename... Strings> void add_challenge(size_t round, Strings&... labels)
+    {
+        manifest[round].challenge_label = { labels... };
+    }
+    void add_entry(size_t round, std::string element_label, size_t element_size)
+    {
+        manifest[round].entries.emplace_back(element_label, element_size);
+    }
+
+    bool operator==(const TranscriptManifest& other) const = default;
+};
+
+/**
+ * @brief Common transcript functionality for both parties. Stores the data for the current round, as well as the
+ * manifest.
+ *
+ * @tparam FF Field from which we sample challenges.
+ */
+template <typename FF> class BaseTranscript {
+    // TODO(Adrian): Make these tweakable
+    static constexpr size_t HASH_OUTPUT_SIZE = 32;
+    static constexpr size_t MIN_BYTES_PER_CHALLENGE = 128 / 8; // 128 bit challenges
+
+    size_t round_number = 0;
+    std::array<uint8_t, HASH_OUTPUT_SIZE> previous_challenge_buffer{};
```

Comment:
Can we explicitly `std::fill` this with `0`s?

Reply:
The  fact that we use the `{}` means that in the present case it is filled with zeros, but I prefer to be explicit rather than to rely on the behavior of the underlying type (I don't know what guarantees exist).

Reply:
Sure, we can do that

Reply:
On second thought, to do this I need to add a constructor where there was not one before:
`  public:
    BaseTranscript ()
    {
        previous_challenge_buffer.fill(0);
    }`

As you pointed out, zero initialization is guaranteed in this case so filling with zeros would really only serve to clarify. I'm not sure this even achieves that purpose but I'll default to your preference.

Reply:
Update: after doing some digging, I'm having a hard time convincing myself beyond any doubt that zero initialization is indeed guaranteed in this case. I asked @adr1anh if he has confidence on this and if he could point us to a reference. We can of course just go with the extra safe option but now I'm just curious.

Reply:
Adrian referenced this [post](https://stackoverflow.com/questions/18295302/default-initialization-of-stdarray). This is also helpful [](https://en.cppreference.com/w/cpp/language/zero_initialization)